### PR TITLE
Ft multi city request 167735045

### PR DESCRIPTION
**Pull request by SamuelKoroh**
**What does this PR do?**

  The PR allow user to request to visit multiple city with just one trip request.

**Description of Task to be completed?**

- create request model.
- create request controller.
- create request service.
- create the route to handle the endpoint.
- document the endpoint in the swagger document.
- create middle-ware to authenticate the user
- write test for the endpoint

 **How should this be manually tested?**

 - run the server and on postman create a post request to http://localhost:3000/api/v1/request/multi-city, passing array of request destination
- from the swagger document go to the Request section there also you can send a post request to http://localhost:3000/api/v1/request/multi-city, passing array of request destination
- provide valid token for a registered user.

 **What are the relevant stories?**

  https://www.pivotaltracker.com/story/show/167735045

 **Screenshots (if appropriate)**
  
image

### CODE REVIEW

File: src/controllers/requestController.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export default class RequestController {
+  /**
+   * @param {req} req - request obiect
+   * @param {res} res - res object object
+   * @returns {object} - return object
+   */
+  static async multiCityRequest(req, res) {
+    try {
+      const data = await RequestServices.multiCityRequest(req);
+      res.status(200).json({ status: 'success', data });
```
Comment (Annmary12) [inline src/controllers/requestController.js]: Check the util folder, there is a helper function for response, please utilize that and adjust all your code

File: src/migrations/20190829130939-create-department.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    },
+    manager: {
+      type: Sequelize.INTEGER,
+      references: {
+        model: 'Users',
+        key: 'id',
+        as: 'manager'
+      }
+    },
+    createdat: {
```
Comment (Annmary12) [inline src/migrations/20190829130939-create-department.js]: This should be `createdAt` not `createdat`, the same goes with updatedat

File: src/migrations/20190829131034-create-profile.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Profiles', {
```
Comment (Annmary12) [inline src/migrations/20190829131034-create-profile.js]: Why do we have `Profile` table where we have the `Users` table? please remove it and add `department_id` as a filed in the Users table.

File: src/migrations/20190829012904-create-multi-request.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('MultiRequests', {
```
Comment (Annmary12) [inline src/migrations/20190829012904-create-multi-request.js]: It does not make sense to have two tables as `request` table(request and multi requests), this means our application is not scalable. We are going to have one table which is `request` table then the destination field will be an array to accept multiple destinations.

File: src/migrations/20190829130939-create-department.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    department: {
+      type: Sequelize.STRING
+    },
+    manager: {
```
Comment (Annmary12) [inline src/migrations/20190829130939-create-department.js]: And why do we have manager field, also why is it tied to `users` table?

File: src/routes/index.js [hunk 6]
```diff
@@ -1,8 +1,11 @@
 import { Router } from 'express';
 import authRoute from './auth';
+import requestRoute from './request';
+import auth from '../middlewares/auth';
 
 const router = Router();
 
 router.use('/auth', authRoute);
+router.use('/request', auth, requestRoute);
```
Comment (Annmary12) [inline src/routes/index.js]: `request` should be plural `requests`

File: src/routes/request.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
+import RequestController from '../controllers/requestController';
+
+const router = express.Router();
+
+router.post('/multi-city', RequestController.multiCityRequest);
```
Comment (Annmary12) [inline src/routes/request.js]: Please apply destructuring in your code. 

Example
Instead of having `RequestController.multiCityRequest`

you can destructure it up by:
```
const { multiCityRequest } = RequestController;

router.post('/multi-city', multiCityRequest);

```

File: src/services/requestService.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Class for managing trip request
+ */
+export default class RequestService {
+  /**
+   * @param {object} body - arrays of request object
+   * @returns {object} obej - return object
+   */
+  static async multiCityRequest({ body, user }) {
+    if (body.length < 1) throw new Error('The request cannot be empty');
```
Comment (Annmary12) [inline src/services/requestService.js]: Instead of doing this ` if (body.length < 1) throw new Error('The request cannot be empty');`, please make use of our validation schema.

File: src/migrations/20190829012904-create-multi-request.js (hunk 4 above)
Comment (SamuelKoroh) [reply src/migrations/20190829012904-create-multi-request.js]: Noted

File: src/migrations/20190829130939-create-department.js (hunk 5 above)
Comment (SamuelKoroh) [reply src/migrations/20190829130939-create-department.js]: A department has a manager and a manager is a user

File: src/controllers/requestController.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  /**
+   * @param {req} req - request obiect
+   * @param {res} res - res object object
+   * @returns {object} - return object
+   */
+  static async multiCityRequest(req, res) {
+    try {
+      const data = await RequestServices.multiCityRequest(req);
+
+      Response.success(res, data);
```
Comment (cleave3) [inline src/controllers/requestController.js]: you can use object destructuring here e.g `const { success, badRequest} = Response`
so you can have `success(res, data, 201)`. You can do this in other places where this applies. I think this will make your code cleaner

File: src/migrations/20190829130939-create-department.js (hunk 5 above)
Comment (Annmary12) [reply src/migrations/20190829130939-create-department.js]: We should have a `Role table` instead of `Department`

File: src/models/User.js [hunk 10]
```diff
@@ -2,17 +2,14 @@ module.exports = (sequelize, DataTypes) => {
...
-      },
       first_name: {
         type: DataTypes.STRING,
         allowNull: false,
 
       },
+      middle_name: {
+        type: DataTypes.STRING
+      },
       last_name: {
```
Comment (Annmary12) [inline src/models/User.js]: Why did you add `middle_name`, did the team approve this?

File: src/utils/response.js [hunk 11]
```diff
@@ -31,9 +42,9 @@ export default class Response {
    * @param {string} status - response status
    * @returns {object} response - success response
    */
-  static contentRemoved(res, message, statusCode = 200, status = 'success') {
-    return res.status(statusCode).json({ status, data: { message } });
-  }
+  // static contentRemoved(res, message, statusCode = 200, status = 'success') {
+  //   return res.status(statusCode).json({ status, data: { message } });
+  // }
```
Comment (Annmary12) [inline src/utils/response.js]: Remove commented out code

File: src/validation/validator.js [hunk 12]
```diff
@@ -10,13 +11,28 @@ const validate = (schema, property) => (req, res, next) => {
...
+
+  if (errors.error) {
+    const errorMSG = [];
+
+    errors.error.details.forEach(err => errorMSG.push(err.message));
+
+    return Response.badRequest(res, errorMSG);
   }
+  next();
 };
```
Comment (Annmary12) [inline src/validation/validator.js:37]: Please, why do we have two validators function in this file? the one @cleave3 created and yours.

Comment (SamuelKoroh) [reply src/validation/validator.js:37]: The first returns just single validation error while the one I create return more than one validation error for now I don't want others code to break that why I didn't touch the one created by @cleave3

File: src/utils/response.js (hunk 11 above)
Comment (SamuelKoroh) [reply src/utils/response.js]: Ok

File: src/models/User.js (hunk 10 above)
Comment (SamuelKoroh) [reply src/models/User.js]: The middle_name was returned by Facebook

File: src/controllers/requestController.js (hunk 9 above)
Comment (SamuelKoroh) [reply src/controllers/requestController.js]: Ok I will do that

File: src/validation/validator.js (hunk 12 above)
Comment (Annmary12) [reply src/validation/validator.js:37]: @SamuelKoroh , we should use one. you can adjust other validations to use the one that has multiple validation. and test login and signup to ensure nothing is breaking

File: src/models/User.js (hunk 10 above)
Comment (Annmary12) [reply src/models/User.js]: Is not compulsory that you have to store that in our database if the team didn't include that.

File: src/validation/validator.js (hunk 12 above)
Comment (SamuelKoroh) [reply src/validation/validator.js:37]: I did that and a lot of test were failing that's why I option to do  it separately.

File: src/models/User.js (hunk 10 above)
Comment (SamuelKoroh) [reply src/models/User.js]: OK,  should I remove it. Because I didn't make it a required field.

File: src/migrations/20190828145652-create-request.js [hunk 13]
```diff
@@ -13,48 +12,48 @@ module.exports = {
...
-      allowNull: true,
+    return_date: {
+      type: Sequelize.DATE
     },
-    tripType: {
-      type: Sequelize.STRING,
+    trip_type: {
+      type: Sequelize.ENUM,
       allowNull: false,
+      values: ['one-way', 'return', 'multi-city'],
```
Comment (Annmary12) [inline src/migrations/20190828145652-create-request.js:28]: Why are you not using camelcase?

File: src/migrations/20190828145652-create-request.js [hunk 14]
```diff
@@ -13,48 +12,48 @@ module.exports = {
...
-    },
-    userId: {
+    user_id: {
       type: Sequelize.INTEGER,
       onDelete: 'CASCADE',
       references: {
         model: 'Users',
         key: 'id',
-        as: 'userId'
+        as: 'user_id'
```
Comment (Annmary12) [inline src/migrations/20190828145652-create-request.js]: Please change all these to camelcase

Comment (SamuelKoroh) [reply src/migrations/20190828145652-create-request.js]: Yes it camelCase

File: src/migrations/20190828145652-create-request.js (hunk 13 above)
Comment (SamuelKoroh) [reply src/migrations/20190828145652-create-request.js:28]: it's just a value not a field
